kernel/sched.c

```diff
@@ -1752,48 +1759,37 @@ void z_thread_abort(struct k_thread *thread)
 	}
 
 #ifdef CONFIG_SMP
-	if (is_aborting(thread) && thread == _current && arch_is_in_isr()) {
-		/* Another CPU is spinning for us, don't deadlock */
-		end_thread(thread);
```

Comment:
Oh... actually this is wrong.  The comment is telling us why: we have a spin condition below where we're going to wait for another thread to die before we get to the end of the code and call end_thread().  But if that thread is itself in the same loop waiting on us to die, it's never going to reach end_thread() either and we'll deadlock.  So this can't be simplified, this has to be here.

I suspect the halt handling likewise has a similar deadlock that needs to be handled (i.e. two threads racing trying to suspend each other).

Reply:
I see what you mean. I think the logic needed here is ...

If the current thread is flagged as aborting, first call end_thread(_current) and then call z_swap() if at thread-level, or continue on if in an ISR.

Reply:
No need to handle the swap case, the IPI will preempt any thread-level logic anyway.  Honestly I think this code was fine as-is?